[HAL] Add configuration stage in executable lowerings

qedawkins opened the pull request:
The core idea behind this split is to allow for passes to be added in between strategy selection and translation, thereby expanding and contracting the set of variants per executable based on the known feature requirements for those strategies.

The reason a split is needed is because strategy selection itself should not care what other variants of the executable are being compiled, however to deduplicate them we need to nest at the executable level. Codegen was already partially structured to enable this (by way of encoding the lowering pipelines as an IR attribute). This PR just recovers the separation that was already there.

In terms of immediate benefits, this adds an option to dump executable configurations (essentially post-kernel config dispatches) and substitute them in, making it significantly easier to get the default lowering configuration for a dispatch from the compiler, tweak it, and plug it back in. Additionally benchmark dumps now happen after this stage for the same reason.

qedawkins committed:
[HAL] Add configuration stage in executable lowerings

The core idea behind this split is to allow for passes to be added in
between strategy selection and translation, thereby expanding and
contracting the set of variants per executable based on the known
feature requirements for those strategies.

The reason a split is needed is because strategy selection itself should
not care what other variants of the executable are being compiled,
however to deduplicate them we need to nest at the executable level.
Codegen was already partially structured to enable this (by way of
encoding the lowering pipelines as an IR attribute). This PR just
recovers the separation that was already there.

In terms of immediate benefits, this adds an option to dump executable
configurations (essentially post-kernel config dispatches) and
substitute them in, making it significantly easier to get the default
lowering configuration for a dispatch from the compiler, tweak it, and
plug it back in. Additionally benchmark dumps now happen after this
stage for the same reason.

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Target/TargetBackend.h:
```diff
@@ -149,6 +154,34 @@ class TargetBackend {
+  // that the pipeline will only run on executable contents.
+  //
+  // The primary purpose of this pipeline is to do the minimum necessary
+  // preprocessing on variants necessary for translation, as well as set any
```
Should remove the duplicated "necessary" here.

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Target/TargetBackend.h:
```diff
@@ -149,6 +154,34 @@ class TargetBackend {
+  //       }
+  //       module { ... }
+  //     }
+  //   }
```
Can we also add doc about potential expected output here?

qedawkins replied:
The doc for this is a little strange. I just tried to indicate that it is structurally the same with some annotations.

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Target/TargetBackend.h:
```diff
@@ -149,6 +154,34 @@ class TargetBackend {
+  // preprocessing on variants necessary for translation, as well as set any
+  // strategy specific information needed for translation. This is to inform
+  // later passes on how to expand/contract variants for multiple targets that
+  // might be known to translate to the same thing. As such, this pipeline is
```
The meaning of "might be known to translate to the same thing" is unclear to me. Could you revise to make it clearer?

qedawkins replied:
Let me know if the wording is clearer now.

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Transforms/ConfigureExecutables.cpp:
```diff
@@ -0,0 +1,165 @@
+// Copyright 2020 The IREE Authors
```
2023

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Transforms/ConfigureExecutables.cpp:
```diff
@@ -0,0 +1,165 @@
+
+    // This pipeline is optional, and the default is no passes, in which case
+    // nothing is needed.
+    if (!passManager.empty()) {
```
Style: prefer early return here to drop one level of nesting for the following.

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Transforms/ConfigureExecutables.cpp:
```diff
@@ -0,0 +1,165 @@
+    // nothing is needed.
+    if (!passManager.empty()) {
+      if (failed(runPipeline(passManager, variantOp))) {
+        variantOp.emitError() << "failed to run translation of source "
```
The error message here needs to be updated?

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.h:
```diff
@@ -31,6 +31,9 @@ enum class PipelinePhase {
   // Runs the transform pipeline up to executable sources (pre translation).
   ExecutableSources,
+  // Runs the transform pipeline up to executable configurations (before
+  // strategy selection).
```
"before translation strategy selection" to be clear and consistent with the following.

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.h:
```diff
@@ -131,6 +134,15 @@ createPreprocessExecutablesWithPipelinePass(std::string pipeline);
 std::unique_ptr<OperationPass<IREE::HAL::ExecutableOp>>
 createPreprocessExecutablesWithToolPass(std::string command);
 
+// Configures hal.executable.variant ops via a nested translation pipeline.
```
Configures hal.executable.variant ops in all hal.executable ops via ..

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Target/TargetBackend.cpp:
```diff
@@ -65,6 +67,13 @@ void TargetOptions::bindOptions(OptionsBinder &binder) {
+      "iree-hal-dump-executable-configurations-to",
+      executableConfigurationsPath,
+      llvm::cl::desc("Path to write individual hal.executable input source "
+                     "listings into, after configuration (- for stdout)."),
```
.. after translation selection and before starting translation

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp:
```diff
@@ -83,6 +83,27 @@ static llvm::cl::opt<std::string> clSubstituteExecutableSourcesFrom{
+static llvm::cl::list<std::string> clSubstituteExecutableConfiguration{
+    "iree-hal-substitute-executable-configuration",
+    llvm::cl::desc(
+        "A `executable_name=object_file.xxx` pair specifying a "
```
This description is copied over from `iree-hal-substitute-executable-object` and needs to be updated?

antiagainst commented on compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp:
```diff
@@ -242,6 +251,55 @@ void buildHALTransformPassPipeline(OpPassManager &passManager,
+  //----------------------------------------------------------------------------
+
+  if (compileFrom < PipelinePhase::ExecutableConfigurations) {
+    // Select a strategy for each hal.executable.variant and generate the IR to
```
".. a translation strategy .." to be clear and relevant to hal. I guess we should also look at other places where we only mention "strategy" by itself and do the same when suitable. (I've commented at a few other places but not exhaustive for sure.)

antiagainst commented on tools/test/executable_configurations.mlir:
```diff
@@ -0,0 +1,44 @@
+// RUN:     --mlir-print-ir-before=iree-hal-serialize-executables 2>&1 | \
+// RUN: FileCheck %s
+
+// This test relies on us piping stdout and that there's only a single
```
s/us/using/

antiagainst reviewed (changes requested):
Cool stuff, thanks Quinn! LGTM overall.

qedawkins committed:
Address comments

 - Add clearer description of the intent behind the pipeline
 - Strategy -> translation strategy where relevant
 - Update cli/error messages

qedawkins committed:
Update configuration pipeline descriptions

qedawkins committed:
strategies -> translation strategies

qedawkins committed:
oops, turn codegen back on
